web-api/src/personalization/filter.rs

```diff
@@ -288,6 +308,74 @@ impl Filter {
             Self::Combine(combine) => combine.filters.is_below_depth(max),
         }
     }
+
+    pub(crate) fn insert_published_after(
+        filter: Option<Self>,
+        published_after: Option<DateTime<Utc>>,
+    ) -> Option<Self> {
+        if let Some(published_after) = published_after {
+            let field = "publication_date".try_into().unwrap(/* valid property id */);
+            let value = DocumentProperty::try_from_value(
+                &"unused".try_into().unwrap(/* valid document id */),
+                &field,
+                json!(published_after.to_rfc3339()),
+            ).unwrap(/* valid property */);
+            let published_after = Self::Compare(Compare {
+                operation: CompareOp::Gte,
+                field,
+                value,
+            });
+
+            let filter = if let Some(filter) = filter {
+                match filter {
+                    compare @ Self::Compare(_) => Self::Combine(Combine {
+                        operation: CombineOp::And,
+                        filters: Filters(vec![compare, published_after]),
+                    }),
+                    Self::Combine(mut combine) if matches!(combine.operation, CombineOp::And) => {
+                        combine.filters.push(published_after);
+                        Self::Combine(combine)
+                    }
+                    combine @ Self::Combine(_) => Self::Combine(Combine {
+                        operation: CombineOp::And,
+                        filters: Filters(vec![combine, published_after]),
+                    }),
+                }
+            } else {
+                published_after
+            };
+
+            Some(filter)
+        } else {
+            filter
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for Filter {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        // this is essentially what #[serde(untagged)] generates, but with better error messages
+        // than the rather inexpressive "data did not match any variant of untagged enum Filter"
+        use serde::__private::de::{Content, ContentRefDeserializer};
```

Comment:
i know this is not very nice, but i tripped multiple times over the issue that `#[serde(untagged)]` generates a pretty useless error message. if there is any better way to bubble up the errors of the variants, then i'm happy to change this.